Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/main.cpp

```diff
@@ -143,43 +125,85 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     StringBuilder output_builder;
 
-    if (namespace_header_mode)
-        IDL::generate_namespace_header(interface, output_builder);
-
-    if (namespace_implementation_mode)
-        IDL::generate_namespace_implementation(interface, output_builder);
-
-    if (constructor_header_mode)
-        IDL::generate_constructor_header(interface, output_builder);
-
-    if (constructor_implementation_mode)
-        IDL::generate_constructor_implementation(interface, output_builder);
-
-    if (prototype_header_mode)
-        IDL::generate_prototype_header(interface, output_builder);
-
-    if (prototype_implementation_mode)
-        IDL::generate_prototype_implementation(interface, output_builder);
-
-    if (iterator_prototype_header_mode)
-        IDL::generate_iterator_prototype_header(interface, output_builder);
+    auto write_file = [&](auto generator_function, StringView file_path) -> ErrorOr<void> {
+        (*generator_function)(interface, output_builder);
+
+        auto output_file = TRY(Core::File::open(file_path, Core::File::OpenMode::ReadWrite));
+
+        // Only write to disk if contents have changed
+        auto previous_contents = TRY(output_file->read_until_eof());
+        TRY(output_file->seek(0, SeekMode::SetPosition));
+        if (previous_contents != output_builder.string_view())
+            TRY(output_file->write_until_depleted(output_builder.string_view().bytes()));
+        // FIXME: Can we add clear_with_capacity to StringBuilder instead of throwing away the allocated buffer?
+        output_builder.clear();
+        return {};
+    };
+
+    String namespace_header;
+    String namespace_implementation;
+    String constructor_header;
+    String constructor_implementation;
+    String prototype_header;
+    String prototype_implementation;
+    String iterator_prototype_header;
+    String iterator_prototype_implementation;
+    String global_mixin_header;
+    String global_mixin_implementation;
+
+    auto path_prefix = LexicalPath::join(output_path, lexical_path.stem());
+
+    if (interface.is_namespace) {
+        namespace_header = TRY(String::formatted("{}Namespace.h", path_prefix));
+        namespace_implementation = TRY(String::formatted("{}Namespace.cpp", path_prefix));
+
+        TRY(write_file(&IDL::generate_namespace_header, namespace_header));
+        TRY(write_file(&IDL::generate_namespace_implementation, namespace_implementation));
+    } else {
+        constructor_header = TRY(String::formatted("{}Constructor.h", path_prefix));
```

Comment:
nit: What I usually do when the computation of the output files is in two places: Pass the list of output files the build system expects to the tool, and make the tool check that it's in fact identical to the list of output files the tool writes. (So that if the naming is changed in one place, we don't forget to update it in the other.)

Reply:
Hmm that's a good idea. I'll throw that one on the "future improvements" yak stack that I pulled this change off of.